### DISCUSSION TITLE
I think I've found a bug (or missing feature) related to query building and detecting model relationships, can I open an issue for this?

### DISCUSSION THREAD

**Discussion by Evertt** [larastan/larastan#1895] [Q&A]
Okay, so I have a model called `TrainingDate`, which has the following `scope` function:

```php
/**
 * @param Builder<static> $query
 */
public function scopeSearch(Builder $query, string $search)
{
    return $query->where(
        /**
         * @param Builder<static> $query
         */
        function (Builder $query) use ($search) {
            // These few lines are just to check whether the search string
            // happens to be a valid date string.
            try {
                $date_search = Carbon::parse($search);
            } catch (\Throwable $e) {
                // If it wasn't, then we'll just define a $date_search
                // that will never match any row, so it will be inconsequential
                $date_search = Carbon::parse('00-00-0000');
            }
            
            $query
                // I have commented the following line out on purpose.
                // If I would re-activate it, then it would make phpstan show errors.
                // ->whereDate('date', $date_search) 
                ->orWhereHas('training', function ($training) use ($search) {
                    $training->where('title', 'like', "%{$search}%");
                })
                ->orWhereHas('training_location', function ($training_location) use ($search) {
                    $training_location->where('title', 'like', "%{$search}%");
                })
                ->orWhereHas('business', function ($business) use ($search) {
                    $business->where('name', 'like', "%{$search}%");
                })
            ;
        }
    );
}
```

So, as the code stands right now, phpstan doesn't show any false errors. If I would intentionally misspell one of the names of the relationships, like turning `'training'` into `'ttraining'`, phpstan will correctly catch that.

However, if I re-activate the line `->whereDate('date', $date_search)`, then suddenly phpstan will show an error for every `orWhereHas` line after `whereDate`, saying that all the relationship names `'business'`, `'training_location'`, and `'training'` do not exist on `Illuminate\Database\Eloquent\Builder model`. That error implies to me that somehow the information about the generic type `Builder<TrainingDate>` was lost somewhere on the way.

Which is slightly confusing, because when I hover over `whereDate()`, it seems to show me a correct return-type of `Builder<TrainingDate>`:

![image](https://github.com/larastan/larastan/assets/1267282/8c6ae4b4-182d-450b-b5d6-578236ed72de)

And `Builder<TrainingDate>` is also the type of `$query`, so if all the `orWhereHas()` statements were working correctly when coming directly from `$query`, then I'd assume it should also work correctly coming after `whereDate()`.

The only thing I can see that fixes this in the short-term, is to replace `whereDate()` with a simple `where()` clause. Because the simple `where()` clause goes to `Illuminate\Database\Eloquent\Builder` instead of `Illuminate\Database\Query\Builder`. Which seems to have no further consequences for my code, so I guess I can do that for now.

But still, I think it would be an improvement if phpstan would understand that once you go down from `Illuminate\Database\Eloquent\Builder<TrainingDate>` to `Illuminate\Database\Query\Builder`, that you can always go back to `Illuminate\Database\Eloquent\Builder` without losing the `<TrainingDate>` information.

### Edit

Here's another example of such an error which doesn't seem to make sense:

![image](https://github.com/larastan/larastan/assets/1267282/bcb06611-c82a-485d-9685-4542a77d4325)

Do you see how weird the return-type is? And again, this only happens because `whereBetween` resides in `Illuminate\Database\Query\Builder`. Every method that comes from `Illuminate\Database\Eloquent\Builder` seems to work fine...

**Comment by calebdw** [comment]
Likely fixed by #1874

**Reply by calebdw** [reply]
Also, **don't** specify the parameter type for closures as this overrides the generics.

Use `function ($query)` instead of `function (Builder $query)`

**Reply by Evertt** [reply]
> Also, **don't** specify the parameter type for closures as this overrides the generics.

It doesn't seem to do that for me?

![image](https://github.com/larastan/larastan/assets/1267282/7c321e98-300c-4d54-882b-8df8b5209718)

### Edit

As long as this is working as it seems to be working for me, I actually prefer it this way. Because this seems a little more backwards compatible, you know. Whatever static analysis tool doesn't understand generics in docblocks will at least be able to understand the very explicit `Builder` type for the parameter.

**Reply by calebdw** [reply]
I'm just saying that I've had issues and have had to remove them.

I suspect that's also the reason for your nested `Builder<Builder<Model>>`

**Reply by Evertt** [reply]
@calebdw Well, I think you may be almost right?

https://github.com/larastan/larastan/assets/1267282/737439db-a263-4e9c-b1f2-a7910c2fd76b

(Turn the sound on...)

**Reply by calebdw** [reply]
So I use explicit custom builder classes which are much more phpstan friendly (less magic going on). But you can try to add the php `Builder` return type and remove the phpstan generic return type

**Comment by canvural** [comment]
Hi,

I tested this with
```php
assertType('Illuminate\Database\Eloquent\Builder<App\User>', User::query()->where(function(Builder $query) {
        assertType('Illuminate\Database\Eloquent\Builder<App\User>', $query->whereDate('name', '2020-01-01'));
    }));
```

which passes. So I cannot reproduce it. If you are still having issues, please create a small repository that can reproduce the issue.
